File: scripts/_cmp.py

const { PNG } = require('pngjs');
const fs = require('fs');

function load(file){
  return PNG.sync.read(fs.readFileSync(file));
}
function analyze(file){
  const p = load(file);
  let rs=0,gs=0,bs=0,n=0;
  // sample every 7th pixel
  for(let y=0;y<p.height;y+=4){
    for(let x=0;x<p.width;x+=4){
      const i=(p.width*y+x)<<2;
      rs+=p.data[i]; gs+=p.data[i+1]; bs+=p.data[i+2]; n++;
    }
  }
  return {w:p.width,h:p.height,r:(rs/n/255).toFixed(3),g:(gs/n/255).toFixed(3),b:(bs/n/255).toFixed(3)};
}
const real=analyze('cmp_real.png');
const live=analyze('cmp_live.png');
console.log('REAL', JSON.stringify(real));
console.log('LIVE', JSON.stringify(live));
// crude similarity: compare avg RGB distance
const dr=Math.abs(real.r-live.r), dg=Math.abs(real.g-live.g), db=Math.abs(real.b-live.b);
console.log('RGB_DELTA', dr.toFixed(3), dg.toFixed(3), db.toFixed(3));

File: scripts/cmp.js
const { chromium } = require('playwright');
const { PNG } = require('pngjs');

(async () => {
  const browser = await chromium.launch();
  const page = await browser.newPage({ viewport: { width: 1280, height: 1000 } });
  await page.route('**/*', (route) => {
    const u = route.request().url();
    if (u.includes('allmylinks.com') || u.includes('beacons.ai') || u.includes('googleusercontent.com/embeds') || u.includes('doubleclick') || u.includes('google-analytics')) {
      return route.abort();
    }
    return route.continue();
  });

  async function shot(url, file, full=false) {
    await page.goto(url, { waitUntil: 'load', timeout: 45000 });
    await page.waitForTimeout(full ? 4000 : 2000);
    await page.screenshot({ path: file, fullPage: full });
  }

  try { await shot('https://www.purecomp.net/', 'cmp_real.png'); console.log('real done'); } catch(e){ console.log('real fail', e.message.split('\n')[0]); }
  try { await shot('https://nate-mina.github.io/puco-py/?v='+Date.now(), 'cmp_live.png'); console.log('live done'); } catch(e){ console.log('live fail', e.message.split('\n')[0]); }

  await browser.close();
})();
